src/pinning/pin-remote-add-rule-builder.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * SPDX-FileCopyrightText: 2022 Tim Perry <tim@httptoolkit.tech>
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as Mockttp from "mockttp";
+import {
+    buildIpfsFixedValueResponse,
+    IpfsFixedResponseHandlerDefinition
+} from "../utils/http";
+
+/**
+ * A builder to allow defining rules that will mock IPFS pin remote add requests.
+ */
+export class PinRemoteAddRuleBuilder {
+
+    /**
+     * This builder should not be constructed directly. Call `mockNode.forPinRemoteAdd(cid)` instead.
+     */
+    constructor(
+        private cid: string | undefined,
+        private addRuleCallback: (data: Mockttp.RequestRuleData) => Promise<void>
+    ) {
+        if (cid) {
+            this.matchers.push(new Mockttp.matchers.QueryMatcher({ arg: cid }));
+        } else {
+            this.matchers.push(new Mockttp.matchers.WildcardMatcher());
+        }
+    }
+
+    private matchers: Mockttp.matchers.RequestMatcher[] = [];
+
+    /**
+     * Return a successful result, as if the IPFS content was pinned successfully.
+     *
+     * This method completes the rule definition, and returns a promise that resolves once the rule is active.
+     */
+    thenPinSuccessfully() {
+        return this.addRuleCallback({
+            matchers: this.matchers,
+            handler: this.cid
+                ? new IpfsFixedResponseHandlerDefinition(200, {
+                    Cid: this.cid
+                })
+                : new Mockttp.requestHandlerDefinitions.CallbackHandlerDefinition((req) => {
+                    const url = new URL(req.url);
+                    const cid = url.searchParams.get('arg');
+                    return buildIpfsFixedValueResponse(200, {
+                        Cid: cid
+                    });
+                })
+        });
+    }
+
+    thenFailWith(err: Error | string) {
+        return this.addRuleCallback({
+            matchers: this.matchers,
+            handler: new Mockttp.requestHandlerDefinitions.CallbackHandlerDefinition((req) => {
+                if (err instanceof Error) {
+                    throw err;
+                }
+                throw new Error(err);
```

Comment:
I don't think this is quite right. If you want to return an error response, you'll want to return the proper format of the HTTP response here. Simply throwing an error instead will just respond instead with a 502 with a plain text error string I think (or something similar). IPFS normally returns structured JSON errors, like so: https://github.com/httptoolkit/mockipfs/blob/f6a7720067d6cc454ea4336d82187246da4b0e1f/src/pinning/pin-rm-rule-builder.ts#L58-L78

If you can reproduce the error, you can check what the normal format is, and then define that explicitly here. If you can't, I'd recommend making a guess, but it's probably a 500 with a Message/Code/Type JSON object, similar to the example above.

Independent of the response format, you _might_ want to change the API - elsewhere most of the rule builders fail in specific pre-defined ways like `thenFailAsMissing`, for common types of failures you might want to simulate. If it is useful to have a freeform error string here instead though, fair enough, there are arguments either way.

Reply:
Good call. Rethinking what I did here, this is probably just me being lazy. It will be much better to return a response.  
I'll do some inspecting of real world requests and figure out what the response looks like here.  I'll have an update early next week.

Reply:
I inspected a response from pinata, setup the response to look similar to that.  In case someone wants to be more specific I left the possibility to use a custom response message.